### PR TITLE
Fixed Encoding Bug...Added GitHub downloader, which downloads a zip ball from git hub (doesn't require git to be installed)

### PR DESCRIPTION
**Pull request by amirrajan**
I added a GitHub downloader that explicitly forms a source url for github repos.  This approach doesn't require git to be installed which will lower the barrier to entry for devs who...well...don't have git installed.  I'm hosting a WarmuP template on GitHub and want to make it as easy as possible for a dev to WarmuP my template (don't want to force them to install git).

usage:
warmup http://github.com/amirrajan/loam FoobarWebsite

I use an open source library called DotNetZip to unzip the source from github (MS-PL license): http://dotnetzip.codeplex.com/

localdeploy.bat helped me with testing locally....thought it would be good to have around

### CODE REVIEW

**Comment by amirrajan** [comment]
Apparently I'm really bad at pull requests....I accidentally merged these two features into the same request.  The comments below are related to the "token replacement working while retaining BOM/Preamble" commit.  I fixed the encoding issue...not sure if it needs to be this involved, but I couldn't find a reliable way to detect the BOM and at the same time do string manipulation....so I had to resort to bit shifting.  I'm adding comments to all the pieces that had to change.  This part warranted some testing...so I went ahead and added a test project and incorporated running the tests into build.bat.

File: BomTestFiles/DynamicBlog.slnfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
```
Comment (amirrajan) [inline BomTestFiles/DynamicBlog.slnfile:0]: Added a directory called BomTestFiles.  This file is a solution file straight from one of my projects.  It contains the UTF8 BOM entry.

File: BomTestFiles/DynamicBlog.slnfile.clone [hunk 2]
```diff
@@ -0,0 +1,26 @@
```
Comment (amirrajan) [inline BomTestFiles/DynamicBlog.slnfile.clone:0]: This is a clone of that same sln file generated by the test runner.

File: BomTestFiles/dev.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
```
Comment (amirrajan) [inline BomTestFiles/dev.yml:0]: This is an original .yml file that does NOT contain a BOM entry.

File: BomTestFiles/dev.yml.clone [hunk 4]
```diff
@@ -0,0 +1,7 @@
```
Comment (amirrajan) [inline BomTestFiles/dev.yml.clone:0]: This is the converted .yml file that the TestHarness generated.

File: build.bat [hunk 5]
```diff
@@ -15,6 +15,13 @@ SET build.config.settings="%DIR%\settings\UppercuT.config"
```
Comment (amirrajan) [inline build.bat:14]: I updated the build script to run the build and run all the tests around BOM.

File: lib/nspec.0.9.58/content/DebuggerShim.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
```
Comment (amirrajan) [inline lib/nspec.0.9.58/content/DebuggerShim.cs:0]: Using NSpec as the SpecRunner.  This is a simple way to debug the tests if you need to.

File: lib/nspec.0.9.58/tools/NSpecRunner.vshost.exe.config [hunk 7]
```diff
@@ -0,0 +1,3 @@
```
Comment (amirrajan) [inline lib/nspec.0.9.58/tools/NSpecRunner.vshost.exe.config:0]: Added by Nuget when I installed NSpec and SpecWatchr

File: warmup.Tests/describe_BytesReplacer.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
```
Comment (amirrajan) [inline warmup.Tests/describe_BytesReplacer.cs:0]: These are the tests from BOM/Preamble and string replacement.  It uses the BomTestFiles.

File: warmup.Tests/warmup.Tests.csproj [hunk 9]
```diff
@@ -0,0 +1,67 @@
```
Comment (amirrajan) [inline warmup.Tests/warmup.Tests.csproj:0]: New test project.

File: warmup/BytesReplacer.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
```
Comment (amirrajan) [inline warmup/BytesReplacer.cs:0]: This is the class that does all the bit shifting.  Url references for where I got the logic to do this stuff.

File: warmup/GitHub.cs [hunk 11]
```diff
@@ -0,0 +1,82 @@
```
Comment (amirrajan) [inline warmup/GitHub.cs:0]: The github downloader from the previous commit.  Allows the application of templates hosted on GitHub without the need for GIT.

File: warmup/TargetDir.cs [hunk 12]
```diff
@@ -65,21 +65,23 @@ void ReplaceTokensInTheFiles(DirectoryInfo point, string name)
```
Comment (amirrajan) [inline warmup/TargetDir.cs:64]: String replacement changed to use BytesReplacer

File: warmup/commands/GenerateWarmup.cs [hunk 13]
```diff
@@ -22,15 +22,17 @@ public void Run(string[] args)
```
Comment (amirrajan) [inline warmup/commands/GenerateWarmup.cs:21]: Exporter logic changed for GitHub zip download.

**Comment by amirrajan** [comment]
If you need to me to split up the pull requests (again).  Let me know and I'll redo them.  The GitHub downloader changes are fairly small compared to the encoding bug.

File: build.bat (hunk 5 above)
Comment (ferventcoder) [inline build.bat:14]: I'll implement a custom build step for this

**Comment by ferventcoder** [comment]
Awesome thanks! I have just a couple notes.  The automated build already creates the gem, so you don't need the local deploy. The other is more an observation. In this case I don't mind the introduction of something big like a testing framework, but usually I like to have a discussion up front beforehand to be sure it aligns with the goals of the project.

**Comment by drusellers** [comment]
I am not sure if we will keep the NSpec or not. since that's not what Rob or I use, but the tests are much appreciated. :) Ideally the zip lib would be merged, I have no problem with it. I have used it in other projects myself.

**Comment by drusellers** [comment]
otherwise, everything looks good.

**Comment by amirrajan** [comment]
I have no problem creating NUnit tests instead of the NSpec tests if you're not comfortable with it...NSpec has a special place in my heart though :-D...if you guys are willing, I'm more than happy to discuss some of the benefits NSpec brings to the table over Skype.  

But again, no big deal if you want NUnit instead, just let me know and I'll convert the tests over.

I'm in the Dallas, Texas (Central Time Zone) and am available during lunch or after 5.

**Comment by amirrajan** [comment]
here is the NSpec website: http://nspec.org

and a podcast with Scott Hanselman on it: http://hanselminutes.com/294/understanding-bdd-and-nspec-with-matt-florence-and-amir-rajan

**Comment by amirrajan** [comment]
Would next Wednesday work for you guys?

**Comment by ferventcoder** [comment]
Wednesday is out for me.

**Comment by amirrajan** [comment]
I'll make time whenever.  Just let me know when you guys are
available.  I can also create another pull request that uses NUnit if
you prefer.  Really enjoying warmup's awesomeness, it's made my code
katas in C# start up very fast.

On Fri, Feb 10, 2012 at 12:47 PM, Rob Reynolds
reply@reply.github.com
wrote:

**Comment by amirrajan** [comment]
Just wanted to let you guys know that I'll be presenting at Dallas Days of .Net (http://www.dallasdayofdotnet.com/) and will talk about warmup along with other tools that help increase developer productivity :-).  It's a decent sized conference (I believe Scott Hanselman is flying down from Seattle to also present).  Thanks again for this gem :-).

**Comment by drusellers** [comment]
done

On Wed, Feb 15, 2012 at 10:03 PM, Amir Rajan <
reply@reply.github.com

**Comment by amirrajan** [comment]
woo hoo!

**Comment by ferventcoder** [comment]
Look at you go Dru Sellers... ;)

**Pull request merged**